broadinstitute/ddp-study-server
# [PEPPER-16] Adds patch for location questions in child & dependent add participant activities

# Pepper-16

## Summary
Adds country/state/province questions to the `ADD_PARTICIPANT_PARENTAL` and `ADD_PARTICIPANT_DEPENDENT` surveys, in addition to activity validations for requiring that the location questions are answered, and have an allowable location for the study.

# Testing Process
## Tested on commits
* [v1.32.2](https://github.com/broadinstitute/ddp-study-server/releases/tag/v1.32.2)
    * _patch should be applied, and not skipped_
* 5ae510bebe0dfd01c5536b3f9fcc13a24f21935c (recent [develop](https://github.com/broadinstitute/ddp-study-server/tree/develop))
    * _patch should be applied, and not skipped_
* [pepper-16-aom-country](https://github.com/broadinstitute/ddp-study-server/tree/pepper-16-aom-country)
    * _results in the patch detecting it is unnecessary, resulting in the message_
      ```
      patch will be skipped.
      TASK:: SingularAOMAddParticipantUpdates completed successfully.
      ```

## Study setup
1. Check out the desired version of the `singular` study [broadinstitute/ddp-study-server](/broadinstitute/ddp-study-server)
2. Rebuild dss-server and studybuilder-cli
3. Invalidate the existing `singular` instance in your local environment, if one exists, and re-deploy the `singular` study
4. Switch to this PR's branch: [pepper-16a-aom-country-patch](https://github.com/broadinstitute/ddp-study-server/tree/pepper-16a-aom-country-patch)
5. Run the `SingularAOMAddParticipantUpdates` task
6. Start the DSS backend
7. Run DDP-Angular `singular` instance
    ```
    npx ng serve ddp-singular
    ```

## Testing the functionality
1. Open a browser and visit `http://localhost:4200` _(adjust for your local environment, if necessary)_
2. Click "Sign Me Up!"
3. Complete the prequal
  * `How old are you?` should be an age > 18
  * `Where do you currently live?` should be `United States` or `Canada`
  * `Do you or your immediate family member have a single ventricle heart defect?` should be `Yes`
4. Create a new Auth0 login
_**result:** Your new account should be registered, and your should arrive at the Dashboard, with 3 options: `Enroll Myself`, `Enroll my child`, and `Enroll my adult dependent`_

### Enroll myself
_this option is not impacted by the changes, and does not need to be checked_

### Enroll my child
* Click "Enroll my child" from the Dashboard
* For `Who in your child's family has single ventricle heart defect?` select `The child being enrolled`
* For `How old is your child?`, enter an age < 16
* A question titled `Where does your child currently live?*` should now be visible
  * See PEPPER-16 for acceptance criteria for the location questions
 
### Enroll my adult dependent
* Click "Enroll my adult dependent" from the Dashboard
* For `Who in your child's family has single ventricle heart defect?` select `The dependant being enrolled`
* For `How old is your dependent?`, enter an age > 18
* A question titled `Where does your dependent currently live?*` should now be visible
  * See PEPPER-16 for acceptance criteria for the location questions
